app/gui/qt/utils/sonic_pi_i18n.cpp

```diff
@@ -0,0 +1,175 @@
+      // ...run through the list of preferred languages
+    std::cout << "Looping through preferred ui languages" << std::endl;
+
+    QString l;
+    for (int i = 0; i < preferred_languages.length(); i += 1) {
+      l = preferred_languages[i];
+      l.replace("-", "_");
+
+      //std::cout << preferred_languages[i].toUtf8().constData() << std::endl;
+      if (available_languages.contains(l)) {
+          return l;
+      }
+    }
+  }
+
+  // Fallback to English
+  this->system_language_available = false;
+  return "en";
+}
+
+QStringList SonicPii18n::findAvailableLanguages() {
+  QStringList languages;
+
+  QString m_langPath = root_path + "/app/gui/qt/lang";
+  //std::cout << m_langPath.toUtf8().constData() << std::endl;
+  QDir dir(m_langPath);
+  QStringList fileNames = dir.entryList(QStringList("sonic-pi_*.qm"));
+
+  for (int i = 0; i < fileNames.size(); ++i) {
+    // extract the language from the filename
+    QString lang;
+    lang = fileNames[i]; // "sonic-pi_pt_BR.qm"
+    lang.truncate(lang.lastIndexOf('.')); // "sonic-pi_pt_BR"
+    lang.remove(0, lang.lastIndexOf("sonic-pi_") + 9); // "pt_BR"
+    //lang.replace("_", "-"); // Replace underscores with dashes so it matches the language codes e.g: "pt-BR"
+    //std::cout << lang.toUtf8().constData() << '\n';
+    languages << lang;
+  }
+  // Add the source language
+  languages << "en_GB";
+  languages.sort();
+  return languages;
+}
+
+bool SonicPii18n::loadTranslations(QString lang) {
+  QString language = lang;
+  bool i18n = false;
+  QCoreApplication* app = QCoreApplication::instance();
+
+  // Remove any previous translations
+  app->removeTranslator(&translator);
+  app->removeTranslator(&qtTranslator);
+
+  std::cout << "Loading translations for " << language.toUtf8().constData() << std::endl;
+
+  i18n = translator.load("sonic-pi_" + language, ":/lang/") || language == "en_GB" || language == "en" || language == "C";
+  if (!i18n) {
+    std::cout << "Error: Failed to load language translation for " << language.toUtf8().constData() << std::endl;
+    language = "en";
+  }
+  app->installTranslator(&translator);
+
+  qtTranslator.load("qt_" + language, QLibraryInfo::location(QLibraryInfo::TranslationsPath));
+  app->installTranslator(&qtTranslator);
+
+  return i18n;
+}
+
+QStringList SonicPii18n::getAvailableLanguages() {
+  return this->available_languages;
+}
+
+std::map<QString, QString> SonicPii18n::getNativeLanguageNameList() {
+  return native_language_names;
+}
+
+QString SonicPii18n::getNativeLanguageName(QString lang) {
+  if (lang == "system_language") {
+    return tr("System language");
+  }
+
+  std::map<QString, QString>::iterator it = native_language_names.find(lang);
+  if(it != native_language_names.end()) {
+    // language found
+    return native_language_names[lang];
+  } else {
+    std::cout << "Warning: Predefined language name not found'" << lang.toUtf8().constData() << "'" << std::endl;
```

Comment:
```suggestion
    std::cout << "Warning: Predefined language name not found: '" << lang.toUtf8().constData() << "'" << std::endl;
```

Reply:
(Otherwise we get this kind of thing):
```
[Debug] Adding language zh-Hans to the combo box
Warning: Predefined language name not found'zh-Hans'
```